### PR TITLE
Call "end" after the "write" pipeline

### LINKED ISSUE #195: Do not call ".end()" in the callback of ".write"
Hey 👋 

I've got a question about the way this library calls the original `.end()` method on the `ClientRequest`. It does so in the callback of `.write()` (given there is data to send so that the callback gets called):

https://github.com/follow-redirects/follow-redirects/blob/13136e95bbe23cabbeaeb74bd0c933aa98dd9b96/index.js#L127-L130

I think I understand the intention here: send this chunk, when done (regardless of whether it's successful or not—because you're not checking the `error` callback argument) mark the request as ready. 

## Question

Can we execute `.write` and `.end` in parallel?

```js
this.write(data, encoding, function () {
  self._ended = true;
});

// This call is moved out of the write's callback.
currentRequest.end(null, null, callback);

this._ending = true;
```

## Context

I'm a maintainer of the API mocking library called Mock Service Worker. We intercept requests in Node.js using the [@mswjs/interceptors](https://github.com/mswjs/interceptors) library, which augments the native `ClientRequest` class. We use class augmentation instead of stubbing it to allow as much of the Node.js default behavior to execute without having to polyfill it. Since we are an API mocking tool, we cannot really write request body chunks until we know there's (or isn't) a request handler for this particular request. That's why we only _collect_ body chunks in the `.write()` method:

https://github.com/mswjs/interceptors/blob/219f9f3cd82f5baf3297f7c6095caa356c148d31/src/interceptors/ClientRequest/NodeClientRequest.ts#L85-L97

We then write those chunks only for passthrough requests when the request handler is known. That is possible to check only in the `.end` method:

https://github.com/mswjs/interceptors/blob/219f9f3cd82f5baf3297f7c6095caa356c148d31/src/interceptors/ClientRequest/NodeClientRequest.ts#L170-L179

Now since `RedirectableRequest` only calls `.end` in the callback of `.write`, when our library is applied the `.end` is never called. That happens because we do not call the `.write` callbacks immediately (not semantic, no chunks are being written just yet so it feels wrong to confirm the result by calling a callback), but instead are referred to the `.end`. Note that this difference is purely an implementation detail which should have no effect on the end-consumer. For them, the expected result is intercepted request and mocked response. 

- The original reported issue: https://github.com/mswjs/msw/issues/1125

## Common usage of `.end`

I've also noticed that commonly the `.write` and `.end` are called in parallel:

```js
const req = http.get()
req.write('hello world')
req.end()
```

That is the way to send request body chunk and end a request in about any example I can find. I've never seen the `write(chunk, () => this.end())` usage before but I understand it must have reasons behind it. 

## What I'd like to know

1. What are the reasons to call `.end` only in the callback of `.write`?
2. What are the implications of calling these two methods in parallel?
3. Do you have any recommendations on how to handle your flow in our "interceptors" library?

Thank you for looking into this.

### PR DESCRIPTION
**Pull request by kettanaito**
- Fixes #195

### CODE REVIEW

**Comment by kettanaito** [comment]
Please allow me to test this version of the library before merging. I will post updates below.

**Comment by kettanaito** [comment]
I can confirm that this change fixes the issue by running automated tests in the [reproduction repository](https://github.com/sbrandwoo/msw-timeout-example):

```
 PASS  src/example.test.js
  ✓ works because post data is not defined (27 ms)
  ✓ does not work because there is a post data (7 ms)
```

🎉

File: index.js [hunk 1]
```diff
@@ -126,8 +126,8 @@ RedirectableRequest.prototype.end = function (data, encoding, callback) {
     var currentRequest = this._currentRequest;
     this.write(data, encoding, function () {
       self._ended = true;
-      currentRequest.end(null, null, callback);
     });
+    currentRequest.end(null, null, callback);
     this._ending = true;
```
Comment (RubenVerborgh) [inline index.js:131]: Perhaps swap those two lines actually?

Comment (RubenVerborgh) [reply index.js:131]: Yes, definitely need to be swapped, or we can write after end in the `callback` (let's add a test).

Comment (kettanaito) [reply index.js:131]: You're talking about swapping `_ending` and `_ended` right?

```js
    this.write(data, encoding, function () {
      this._ending = true;
    });
    self._ended = true;
    currentRequest.end(null, null, callback);
```

This makes sense. I think I've overlooked that.

Comment (kettanaito) [reply index.js]: @RubenVerborgh, I've swapped the order of ending/ended state.

**Comment by kettanaito** [comment]
Hey, @RubenVerborgh. Could you please take a look at these changes once again?

File: index.js (hunk 1 above)
Comment (kettanaito) [reply index.js]: > or we can write after end in the callback (let's add a test).

Not sure what you mean here. We can't call `write` in the callback of `end`, if that's what you're referring to. That'd throw an exception (cannot write, already ended).

**Comment by tuanggolt** [comment]
hợp nhất lại các nhánh

**Comment by RubenVerborgh** [comment]
#195 has been closed, so this PR is no longer needed

**Pull request closed without merging**